mlir/include/mlir/Dialect/Rock/Tuning/GridwiseGemmParams.h

```diff
@@ -206,121 +216,154 @@ template <typename InitParamType> class BasePopulateParams {
-                           const InitParamsXDL &params, int64_t &gemmKBlocks,
+                           const InitParamsAccel &params, int64_t &gemmKBlocks,
                            uint32_t numCu);
-
-  LogicalResult isValidBlockwiseGemmXDLOPS(const InitParamsXDL &param,
-                                           Type dataTypeA, Type dataTypeB,
-                                           StringRef arch, uint32_t blockSize);
-
-  LogicalResult populateDerived(const InitParamsXDL &validParams,
+  LogicalResult populateDerived(const InitParamsAccel &validParams,
                                 const PopulateParamsInfo &info,
                                 GemmSize &gemmSize, uint32_t &blockSize,
                                 uint32_t &gridSize, int64_t &gemmKBlocks);
-
   LogicalResult populatePaddingKernelDerived(RockGemmWrapperInterface op,
-                                             const InitParamsXDL &validParams,
+                                             const InitParamsAccel &validParams,
                                              GemmSize &gemmSize,
                                              uint32_t &blockSize,
                                              uint32_t &gridSize);
+};
 
-public:
-  LogicalResult obtainTuningParameters(RockGemmWrapperInterface op,
-                                       uint32_t blockSizeOverride,
-                                       const std::string &perfConfig,
-                                       InitParamsXDL &validParams,
-                                       uint32_t &blockSize, uint32_t &gridSize,
-                                       int64_t &gemmKBlocks);
+//
+// Xdlops interface
+//
+class PopulateParamsXDL : public PopulateParamsAccel {
+  static constexpr size_t nInitParameters = 5;
+  // Initial tuning parameters for forward convolution and backward
+  // convolution.
+  static const InitParamsAccel initParameters[nInitParameters];
 
-  LogicalResult obtainTuningParameters(const PopulateParamsInfo &info,
-                                       uint32_t blockSizeOverride,
-                                       const std::string &perfConfig,
-                                       InitParamsXDL &validParams,
-                                       uint32_t &blockSize, uint32_t &gridSize,
-                                       int64_t &gemmKBlocks);
+  static constexpr size_t nInitParametersFp16 = 3;
+  // Tuning parameters for fp16/bf16 convolutions.
+  static const InitParamsAccel initParametersFp16[nInitParametersFp16];
 
-  std::vector<InitParamsXDL> getTuningParameters(KernelType opType,
-                                                 Type dataTypeA, Type dataTypeB,
-                                                 StringRef arch) const;
+  static constexpr size_t nInitParametersForward8Bit = 5;
+  // Tuning parameters for i8 convolutions.
+  static const InitParamsAccel
+      initParametersForward8Bit[nInitParametersForward8Bit];
```

Comment:
Because that's how I did it back in the day. I think there's some sort of `std::array` stuff you can use to infer the size. I also think there's an issue where you can't have a variable-sized static constant in a class like that because the compiler doesn't know how much `.rodata` it needs for it.

Reply:
Yes, the code was already like this originally. I thought the limitation was also that this would be serialized, used by static libraries, and this is why we wanted a plain array instead of something fancier. Should I try something else?

Reply:
@jerryyin @krzysz00 any further thoughts? I guess this is the last standing point to get this merged?